### PR TITLE
Allow removing HTTP credentials using credentials_override

### PR DESCRIPTION
**Pull request by gferon**
Using an enum like this avoids us to have types like `Option<Option<HttpCredentials>>` that are hard to parse by humans.

This is also helpful for structs that perform authenticated and non-authenticated requests, which can now be a hidden implementation detail.

### CODE REVIEW

File: libsignal-service/src/push_service.rs [hunk 1]
```diff
@@ -119,9 +119,9 @@ pub struct PreKeyStatus {
 }
 
 #[derive(Clone)]
-pub struct HttpCredentials {
-    pub username: String,
-    pub password: String,
+pub enum HttpAuth {
```
Comment (rubdos) [inline libsignal-service/src/push_service.rs:122]: May I suggest a manual implementation of `Debug` here, that prints *only the username* (and possibly "password: [..]" or something like that)?

File: libsignal-service/src/groups_v2/manager.rs [hunk 2]
```diff
@@ -211,13 +211,13 @@ impl<'a, S: PushService, C: CredentialsCache> GroupsManager<'a, S, C> {
...
 
-        Ok(HttpCredentials { username, password })
+        Ok(HttpAuth::Basic { username, password })
     }
 
     pub async fn get_group(
         &mut self,
         group_secret_params: GroupSecretParams,
-        credentials: HttpCredentials,
+        credentials: HttpAuth,
```
Comment (rubdos) [inline libsignal-service/src/groups_v2/manager.rs:220]: What's the semantics of setting this to `HttpAuth::None`?

Comment (gferon) [reply libsignal-service/src/groups_v2/manager.rs:220]: see for example [the first use of it here](https://github.com/Michael-F-Bryan/libsignal-service-rs/pull/85/files#diff-78c1e3016616476335a7b4df6a1171673c2af645b0740309febf03a96f35f95aR367)

I suspect that it'll also be useful for structs that are initialized without credentials and need them later (mostly when linking a secondary device for example).

**Comment by rubdos** [comment]
So, we're modeling tri-state here, right?

1. Do nothing
2. Override default credentials and set *no credentials*
3. Override default credentials and set *these credentials*.

The problem, I think, is that certain functions (notably `put_json` et al) require this tri-state argument, while other methods (e.g. related to the group stuff) require a filed credential. Is it, maybe, an option to have something like

```rust
enum HttpAuthOverride {
    NoOverride,
    OverrideWith(Option<HttpAuth>),
}

struct HttpAuth {}
```

... possibly you can model the full tri-state in `HttpAuthOverride`, getting rid of the `Option`:

```rust
enum HttpAuthOverride {
    NoOverride,
    Unidentified,
    Identified(HttpAuth),
}
```

File: libsignal-service/src/groups_v2/manager.rs (hunk 2 above)
Comment (rubdos) [reply libsignal-service/src/groups_v2/manager.rs:220]: I mean the semantics in the `get_group` specifically.

**Comment by gferon** [comment]
I see, I missed the `get_group` use somehow, which is rather a basic one 😄 I'll try with this suggestion and see how it goes.

File: libsignal-service/src/push_service.rs [hunk 3]
```diff
@@ -112,25 +112,38 @@ pub struct DeviceCapabilities {
...
     pub password: String,
 }
 
-impl ProfileKey {
-    pub fn derive_access_key(&self) -> Vec<u8> {
-        let key = GenericArray::from_slice(&self.0);
-        let cipher = Aes256Gcm::new(key);
-        let nonce = GenericArray::from_slice(&[0u8; 12]);
-        let buf = [0u8; 16];
-        cipher.encrypt(nonce, &buf[..]).unwrap()
```
Comment (rubdos) [inline libsignal-service/src/push_service.rs:133]: Something weird happened to the diff here, mind fixing that up?

**Review by rubdos — CHANGES_REQUESTED** [review]
I like the feel of this, good job.  You like it too?

Small remark wrt. the diff, but it probably doesn't matter. Feel free to decide on that question and merge.

File: libsignal-service/src/push_service.rs (hunk 3 above)
Comment (gferon) [reply libsignal-service/src/push_service.rs]: Done!
